### DISCUSSION TITLE
Fully automated AA with PyTorch

### DISCUSSION THREAD

**Discussion by wangzcl** [aleixalcacer/archetypes#49] [Ideas]
Now we have softmax function to approximately "project" arrays on to the unit simplex, and there is also an entropic descent method which use softmax perform optimization on the unit simplex (https://ieeexplore.ieee.org/abstract/document/10213413).

However, I still feel there is a more straightforward way to use softmax, that is, integrating it into the loss function like what people do to many neural networks. Define the loss function as 

$f=\lVert X - \sigma(A')\sigma(B')X\rVert$, 

where $\sigma$ is the row-wise softmax function. So $A = \sigma(A'), B = \sigma(B')$, and we only need to optimize $A'$ and $B'$. 

What we need to do is just defining the loss function in PyTorch, creating an optimizer object with $A$ and $B$, and PyTorch can do everything else. This would enable us to use some adavanced optimizers like SGD and adam to avoid local minima, and optimize both $A$ and $B$ simultaneously and automatically. It could be especially suitable for large datasets.

**Comment by aleixalcacer** [comment]
If I’m not mistaken, you’re suggesting introducing the softmax inside the loss function. Right?

Feel free to create a PR and see if it works :) 

What do you think?

**Reply by wangzcl** [reply]
Yes. I would try that in the near future.

**Comment by aleixalcacer** [comment]
On the other hand, I should update the numpy biaa to the new API. Right?

If so, I’ll create a PR and would appreciate it if you could review it.

**Reply by wangzcl** [reply]
Happy to do so :)

**Comment by aleixalcacer** [comment]
Another thing: When the library is mature, I was thinking of writing an article about it and submitting it to a journal. Would you like to join?

**Reply by wangzcl** [reply]
That would be awesome. If you need anything that I can help with, just let me know

**Comment by wangzcl** [comment]
I'm working this. So far it has not been as accurate as the simplex PGD one, but using second order optimizer like L-BFGS may fix the issue. Hope no one need to write any more gradient descent manually.

**Reply by aleixalcacer** [reply]
As you may have noticed, I removed the torch and jax modules from the package since the implementation wasn’t the best. However, I have no problem adding them back as long as they’re done properly — as you said, automatic gradient computation is always welcome. It would be nice if, by default, the package only installed the NumPy part, and if archetypes[torch] is specified, then it installs the torch part. That way, we avoid installing PyTorch/JAX/whatever by default.

**Reply by wangzcl** [reply]
I totally agree with that. Perhaps if we are able to do a better one, maybe we should consider making it a seperate package.

**Reply by aleixalcacer** [reply]
In principle, it’s fine to keep everything together. Once you have it a bit more developed, I can help you review it or assist with anything else you need. Another tool you might consider using is Autograd: https://github.com/HIPS/autograd

**Comment by nikifori** [comment]
Hi everyone, I really appreciate your contribution. I would like to ask in what stage is the pytorch counterpart of the archetypal analysis? I am currently running experiments using numpy as backend, but I really think of the pytorch counterpart since in CPU, my experiments takes around 40 hours.

**Reply by wangzcl** [reply]
The Pytorch implementation is exactly what I have been thinking about in the last few days. In principle, we just need to find a smooth mapping from $\mathbb{R}^n$ to unit simplex $\Delta^{n-1}$ so that we can reparameterize the problem as an unconstrained one. It is acutually easier than it might seem, and I may be able to provide a preliminary implementation soon, probably next week.

**Reply by wangzcl** [reply]
And softmax is not the best mapping for this task as it does not provide sparse outputs

**Reply by nikifori** [reply]
What if we translate everything we can in pytorch and the part of the missing mapping be as it is now ? Not sure if it makes sense.

**Reply by wangzcl** [reply]
We have to implement the unit simplex constraint whatsoever. The current implementations (projection, NNLS with penalty) work on CPU, but are difficult to migrate to GPU (you have to write manual backpropogation and others). Migrating other parts of the algorithm to pytorch may not be worth it as simplex constraint is the main computational bottleneck.

One way to implement simplex projection on GPU is [deep-spin/entmax](https://github.com/deep-spin/entmax). However it is known to be slow and I think I might have a better solution than this one. I might have some availability in the next few days to try it out. I'll let you know if there are any updates.

**Comment by wangzcl** [comment]
@nikifori I have implemented a experimental Pytorch AA in my fork: https://github.com/wangzcl/archetypes/tree/torchAA
I did some very basic test and it seem to work, and the result is almost as precise as the numpy PGD version, but I have no idea about its performance on large datasets. APIs subject to change, and I'm happy to hear about your thoughts if you would like to have a try :)

**Reply by aleixalcacer** [reply]
If the PyTorch version offers a substantial performance improvement, I agree to integrate it into the repository. Is there any performance comparison available?

**Reply by wangzcl** [reply]
I haven't done that yet. Do you have any benchmark datasets to recommend?

**Reply by nikifori** [reply]
Hi there, I tested the torch implementation using the following:

<img width="526" height="245" alt="image" src="https://github.com/user-attachments/assets/b9c99ddc-9d81-4125-9f9c-d8b1990f0ebc" />  

`torch_archetypal_analysis_fn` is the `archetypal_analysis` function. Also, for the result photos the `torch_epochs` was 500 and not 2000.

From 45 hours, it took 14 minutes.

| Before (numpy) | After (torch)|
|--------|-------|
| <img width="960" height="720" alt="elbow_n_archetypes" src="https://github.com/user-attachments/assets/8fa3de9a-4dc7-4211-ae7c-564168fee2ba" />| <img width="960" height="720" alt="elbow_n_archetypes" src="https://github.com/user-attachments/assets/bfad1844-4a7d-45c8-a24b-98b77a5dc1d1" />|

There is a slight difference on the RSS metric across the tested number of archetypes. I could probably use more `torch_epochs`.

**Reply by nikifori** [reply]
@wangzcl Hi there,

could you explain the idea of using two nn.Linear instances ?
Is it just a linear layer taking us from data to archetypes and backwards (reconstruction), conditioning the weights to sum into 1 ?
What is also the value of `reproject_cycle` ? Should it be `None` (as it is default or not?)

**Reply by wangzcl** [reply]
Hi @nikifori , sorry for missing your comment last month. Glad to see that the pytorch version seems to work -- maybe with some hyperparameter tuning on the optimizer it will work even better.

The idea is actually simple: The forward expression of AA is $XAB$ (here $X$ is a transposed data batch). Meanwhile, a linear layer without bias or activation is just a linear transformation $Y = XA$, so translating AA into linear layers is just a fancy, pytorch-native way of doing matrix multiplication.

The problem is that, $A$ and $B$ are not just arbitrary matrices. They have to statisfy the unit simplex constriant (say $\Delta^{n-1}$). This is the only thing that hinders a autogradable Pytorch implementation. 

The trick here is to find a differentiable, surjective function $A = f(A'),\\ f:\mathbb{R}^n \to \Delta^{n-1}$ to parameterize A and B (https://github.com/wangzcl/archetypes/blob/74cac110a660ac238c2df106acc4147d97eaa9ca/archetypes/torch/_coreutils.py#L13) and optimize parameter $A'$ directly instead of $A$. If there is an $A^{\*}$ that minimize the objective function and our optimizer is good enough, we will be able to find an $A’^\*$ such that $f(A'^*)=A^\*$.

You can find a more detailed explanation of this technique (called trivialization) in [this paper](https://proceedings.neurips.cc/paper_files/paper/2019/hash/1b33d16fc562464579b7199ca3114982-Abstract.html) and about its [Pytorch integration](https://docs.pytorch.org/tutorials/intermediate/parametrizations.html).

The idea of [reprojection](https://github.com/lezcano/geotorch/issues/48) comes from @lezcano. The gradient of my trivialization function has no radial component (which can be seen by looking at only one row of $A$ and noticing that $f(k A') = f(A')$ ), so $A$ moves along the tangent of the sphere at each gradient descent step, getting further and further from the origin. If it goes to far away, the gradient descent may be less efficient, so it may be helpful to reproject the point onto the unit sphere. I actually haven't tested the feature myself and I don't know how much it can be helpful, so I set the default to `None`.
